File: packages/agentwiki-launcher/test/manifest.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";

import { parseManifest } from "../src/manifest.ts";

function valid() {
  return {
    manifest_version: 1 as const,
    id: "claude-code",
    name: "x",
    tagline: "x",
    icon_url: "/x.svg",
    kind: "local_cli" as const,
    cli_check: { binary: "claude" },
    mcp_config_format: "claude_json" as const,
    first_turn_prompt_delivery: { method: "stdin" as const },
    launch: {
      binary: "claude",
      argv: ["--mcp-config", "${mcp_config_path}"],
      env: { AGENTWIKI_SESSION_ID: "${session_id}" },
      cwd: "${working_dir}",
    },
  };
}

test("valid manifest parses", () => {
  const m = parseManifest(valid());
  assert.equal(m.id, "claude-code");
});

test("unknown var rejected", () => {
  const bad = valid();
  bad.launch.argv.push("${not_a_var}");
  assert.throws(() => parseManifest(bad), /unknown interpolation var/);
});

test("token in argv rejected", () => {
  const bad = valid();
  bad.launch.argv.push("Bearer ${token}");
  assert.throws(() => parseManifest(bad), /token.*forbidden/);
});

test("first_turn_prompt anywhere rejected", () => {
  const bad = valid();
  bad.launch.argv.push("${first_turn_prompt}");
  assert.throws(() => parseManifest(bad), /first_turn_prompt.*forbidden/);
});

test("prompt_file_path in resume rejected", () => {
  const bad = valid() as ReturnType<typeof valid> & {
    resume: { binary: string; argv: string[]; env: Record<string, string> };
  };
  bad.resume = {
    binary: "claude",
    argv: ["--resume", "${prompt_file_path}"],
    env: {},
  };
  assert.throws(
    () => parseManifest(bad),
    /prompt_file_path.*forbidden in resume/,
  );
});

test("unknown manifest version rejected", () => {
  const bad = valid() as { manifest_version: number };
  bad.manifest_version = 2;
  assert.throws(() => parseManifest(bad), /unsupported manifest_version/);
});

File: packages/agentwiki-launcher-go/internal/manifest/manifest_test.go
package manifest

import (
	"encoding/json"
	"strings"
	"testing"
)

func base() map[string]any {
	return map[string]any{
		"manifest_version":    1,
		"id":                  "x",
		"name":                "X",
		"tagline":             "t",
		"icon_url":            "/i.svg",
		"kind":                "local_cli",
		"mcp_config_format":   "claude_json",
		"first_turn_prompt_delivery": map[string]any{"method": "positional_arg"},
		"launch": map[string]any{
			"binary":                "claude",
			"argv":                  []string{"--mcp-config", "${mcp_config_path}"},
			"env":                   map[string]string{"S": "${session_id}"},
			"cwd":                   "${working_dir}",
			"unscoped_workdir_argv": []string{"--permission-mode", "bypassPermissions"},
		},
	}
}

func parseBytes(t *testing.T, m map[string]any) (*Manifest, error) {
	t.Helper()
	b, _ := json.Marshal(m)
	return Parse(b)
}

func TestParseAcceptsValid(t *testing.T) {
	m, err := parseBytes(t, base())
	if err != nil {
		t.Fatalf("Parse: %v", err)
	}
	if m.ID != "x" || m.Launch.Binary != "claude" {
		t.Errorf("unexpected: %+v", m)
	}
}

func TestParseRejectsTokenInArgv(t *testing.T) {
	m := base()
	m["launch"].(map[string]any)["argv"] = []string{"--token", "${token}"}
	_, err := parseBytes(t, m)
	if err == nil || !strings.Contains(err.Error(), "${token}") {
		t.Errorf("expected ${token} rejection, got %v", err)
	}
}

func TestParseRejectsFirstTurnPromptToken(t *testing.T) {
	m := base()
	m["launch"].(map[string]any)["argv"] = []string{"${first_turn_prompt}"}
	_, err := parseBytes(t, m)
	if err == nil || !strings.Contains(err.Error(), "${first_turn_prompt}") {
		t.Errorf("expected first_turn_prompt rejection, got %v", err)
	}
}

func TestParseRejectsInterpolationInUnscopedArgv(t *testing.T) {
	m := base()
	m["launch"].(map[string]any)["unscoped_workdir_argv"] = []string{"${working_dir}"}
	_, err := parseBytes(t, m)
	if err == nil || !strings.Contains(err.Error(), "unscoped_workdir_argv") {
		t.Errorf("expected unscoped_workdir_argv rejection, got %v", err)
	}
}

func TestParseRejectsUnknownVar(t *testing.T) {
	m := base()
	m["launch"].(map[string]any)["argv"] = []string{"${not_a_real_var}"}
	_, err := parseBytes(t, m)
	if err == nil {
		t.Error("expected unknown-var rejection")
	}
}

